src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php

```diff
@@ -12,198 +12,331 @@
+            ->with($this->equalTo($token))
+        ;
+
+        $this->accessDecisionManager
+            ->expects($this->once())
+            ->method('decide')
+            ->with($this->equalTo($token), $this->equalTo(array('foo' => 'bar')), $this->equalTo($this->request))
+            ->will($this->returnValue(true))
+        ;
+
+        $listener = new AccessListener(
+            $this->tokenStorage,
+            $this->accessDecisionManager,
+            $this->accessMap,
+            $this->authenticationManager,
+            $authorizationChecker
+        );
+
+        $this->event
+            ->expects($this->any())
+            ->method('getRequest')
+            ->will($this->returnValue($this->request))
+        ;
+
+        $listener->handle($this->event);
+    }
+
+    public function testHandleWithAuthorizationCheckerWhenAlwaysAuthenticateIsDisabledOnAuthenticatedToken()
+    {
+        $token = new ConcreteToken('user', array('ROLE_FOO'));
+
+        $authorizationChecker = new AuthorizationChecker(
+            $this->tokenStorage,
+            $this->authenticationManager,
+            $this->accessDecisionManager,
+            false
+        );
+
+        $this->tokenStorage
+            ->expects($this->once())
+            ->method('getToken')
+            ->will($this->returnValue($token))
+        ;
+
+        $this->accessMap
+            ->expects($this->any())
+            ->method('getPatterns')
+            ->with($this->equalTo($this->request))
+            ->will($this->returnValue(array(array('foo' => 'bar'), null)))
+        ;
+
+        $this->authenticationManager
+            ->expects($this->never())
+            ->method('authenticate')
+        ;
+
+        $this->tokenStorage
+            ->expects($this->never())
+            ->method('setToken')
+        ;
+
+        $this->accessDecisionManager
+            ->expects($this->once())
+            ->method('decide')
+            ->with($this->equalTo($token), $this->equalTo(array('foo' => 'bar')), $this->equalTo($this->request))
+            ->will($this->returnValue(true))
+        ;
+
+        $listener = new AccessListener(
+            $this->tokenStorage,
+            $this->accessDecisionManager,
+            $this->accessMap,
+            $this->authenticationManager,
+            $authorizationChecker
         );
 
-        $event = $this->getMockBuilder('Symfony\Component\HttpKernel\Event\GetResponseEvent')->disableOriginalConstructor()->getMock();
+        $this->event
+            ->expects($this->any())
+            ->method('getRequest')
+            ->will($this->returnValue($this->request))
+        ;
+
+        $listener->handle($this->event);
+    }
+}
 
-        $listener->handle($event);
+class ConcreteToken extends AbstractToken
+{
+    public function __construct($user, array $roles = array())
+    {
+        parent::__construct($roles);
+        $this->setUser($user);
+        parent::setAuthenticated(count($roles) > 0);
```

Comment:
Is this really needed? Can just set it to true I suppose. It looks like this logic was duplicated from the UsernamePasswordToken, but doesn't exactly serve a purpose. You can also be authenticated without roles.

Reply:
Indeed I wanted to mimic `UsernamePasswordToken`, but `UsernamePasswordToken` has a lot of parameters in it's constructor (lazy)